### PR TITLE
Edges - Kat Perreira

### PR DESCRIPTION
**Pull request by kat-perreira**
# Grocery Store
Congratulations! You're submitting your assignment.

## Comprehension Questions

Question | Response
--- | ---
What is accomplished with `raise ArgumentError`? | Raising an argument error with a message let's the person who is entering in information know specifically what they did wrong and why they needed to enter another thing, like a different status or a different product name.
Why do you think we made the `.all` & `.find` methods class methods?  Why not instance methods? | They are class methods because they are called on all instances of the class. Its easier to call .all on every single creation of an object from the class in one go instead of having to do it after they are all created. 
Think about the relation between `Order` and `Customer`. Is this relation one-to-one, one-to-many, or something else? How does that compare to the Solar System project? | I think it's one to many...one customer has many orders. In the order.csv, some customers made multiple orders as shown by the repeats of customer ids. It's similar but opposite to solar system, where one solar system has many planets. So maybe it's something else.
How is the relation between `Order` and `Customer` tracked in the CSV file? How is it tracked in your program? Why might these be different? | In the csv files, each order has it's own ID number and each customer has it's own ID number. In my program, I mainly am able to search for things by the order number and not the customer ID number of each order that is made. I'm guessing that since this is a grocery store, it's more important to know what the order is and be able to just find who it went to as opposed to how many times a customer ordered something.
Did the presence of automated tests change the way you thought about the problem? How? | I wrote the code to pass the tests. So I started by thinking of the answer I wanted then manipulated my code (a lot) to fix the problem.
CS Fun: If there are `C` customers listed in `customers.csv` and `R` orders listed in `orders.csv`, what is the time complexity of calling `Customer.find`? What is the time complexity of `Order.find`? | I don't thing Shruti finished teaching us this yet but i'll give it a shot...so in customers.find, its searching thru the array once and needs to look over each one with a constant C to compare so its linear which means O(n). And in my Order.find method,  it searched thru the orders array then went thru each of those items in the array to look for the first thing in that item so I think that would be O(n*n) or quadratic.

### CODE REVIEW

**Comment by kat-perreira** [comment]
** I submitted my pull request past my bed time and after reading my answer to the last csFun question, I see that it might be read as passing the blame for my lack of understanding how to write an actual formula and I didn’t mean it like that. Really, I just choose  finish this project first before doing my cs fundamentals homework and the blame for lack of refinement in my solution  lays completely with me

**Comment by tildeee** [comment]
# Grocery Store
## What We're Looking For

Feature | Feedback
--- | ---
**Baseline** |
Answered comprehension questions | x
Used Git Regularly | x, I would prefer commit messages that described code changes/additions/deletions over "wave X complete", but good job otherwise
**Wave 1** |
All provided tests pass | x, actually this code didn't run until i put in a missing `end` in `order.rb`
Using the appropriate `attr_` for instance variables | x
**Wave 2** |
All stubbed tests are implemented fully and pass | actually this code didn't run until i put in a missing `end` in `order.rb`, failing/broken tests for `Order#find`
Used `CSV` library only in `.all` (not in `.find`) | x
Appropriately parses the product data from CSV file in `Order.all` | x
`Order.all` calls `Customer.find` to set up the composition relation | x
**Additional Notes** | 

Good work on this project, Kat!

I think you self-identified that there are places where the code could have been more elegant. I'm writing a small comment on places where I think that's a little more subtle.

Most importantly, running `rake` didn't work until I fixed a syntax error that required another `end` to be put into `order.rb`.

Similarly, I think one of your tests is broken because it's not written correctly, and your `Order#self.find` method isn't working/tested as I'd expect. I'm writing a comment on this!

That being said, the project code is good, meets most of the requirements, and hits the rest of the learning goals.

Good work!

File: lib/customer.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    end
+    return customers
+  end
+
+  def self.find(id)
+    self.all.select do |customer_array|
+      return customer_array if customer_array.id == id
+    end
+    return nil
+  end
```
Comment (tildeee) [inline lib/customer.rb:29]: In this method, `self.all` returns an array of instances of `Customer`. Then, you are using the `Enumerable` method `.select`, which iterates through the array, and you declare each item in the array with the name `customer_array`... which is kind of a weird name for an item in this array of customers?

File: lib/customer.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    @id = id
+    @email = email
+    @address = address
+  end
+
+  def self.all
+    customers = CSV.read('./data/customers.csv').map do |customer_array|
+      id = customer_array[0].to_i
+      email = customer_array[1]
+      address = customer_array[2..5].join(", ")
```
Comment (tildeee) [inline lib/customer.rb:18]: Nice use of ranges and `.join`!

File: lib/order.rb [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    return orders
+  end
+
+  def self.find(order_id)
+    self.all.select do |orders_array|
+      orders_array.each do |item|
+      return orders_array if orders_array[0] == order_id
+    end
+    return nil
+  end
```
Comment (tildeee) [inline lib/order.rb:70]: In this method, `self.all` returns an array of instances of `Order`. Then, you are using the `Enumerable` method `.select`, which iterates through the array, and you declare each item in the array with the name `orders_array`...

Then you iterate through each item of the array (which represents an instance of `Order`, but is named `orders_array`), which doesn't work! You can't call `.each` on an instance of `Order`! This ends up being super strange and broken when we run the tests.

Lastly, if you want to iterate through a collection and define your own conditions for returning, you probably don't want `.select` but want a different way to iterate, like `.each`.

Let me know if you have questions

File: specs/order_spec.rb [hunk 4]
```diff
@@ -141,21 +146,46 @@
...
 
-    it "Returns nil for an order that doesn't exist" do
-      # TODO: Your test code here!
-    end
+  it "Can find the last order from the CSV" do
+    # TODO: Your test code here!
+    final = Customer.find(100)
+
+    expect(final).must_be_kind_of Customer
+    expect(final.id).must_equal 100
```
Comment (tildeee) [inline specs/order_spec.rb:184]: Should this test that's testing "Can find the last order from the CSV" be testing `Customer.find`?

**Pull request closed without merging**